### ISSUE TITLE
`formatlist` broken when interpolating string with `data` because return type is not a `list`

### ISSUE THREAD

**Issue by rkoval** [hashicorp/terraform#16255]
Hi there,

I have an issue regarding `formatlist` and it not properly interpolating strings with `data` such that the resultant type is not actually a `list`.

### Terraform Version
```
Terraform v0.10.6
```

### Terraform Configuration Files
Broken using `data`:

```hcl
provider "aws" {}

data "aws_region" "current" {
  current = true
}

variable "test" {}

resource "aws_rds_cluster" "default" {
  engine             = "aurora"
  cluster_identifier = "asdfasdfasdfa-db-cluster"


  # this line !
  availability_zones = "${formatlist("${data.aws_region.current.name}%s", list("a","b","c"))}"
  # this line !


  master_username    = "${data.aws_region.current.name}" # note how this is also interpolated to demonstrate the issue is only in the `formatlist` context
  master_password    = "asdf"
}
```

Working using `var`:

```hcl
provider "aws" {}

data "aws_region" "current" {
  current = true
}

variable "test" {}

resource "aws_rds_cluster" "default" {
  engine             = "aurora"
  cluster_identifier = "asdfasdfasdfa-db-cluster"


  # this line !
  availability_zones = "${formatlist("${var.test}%s", list("a","b","c"))}" # works now :(
  # this line !


  master_username    = "${data.aws_region.current.name}" # note how this is also interpolated to demonstrate the issue is only in the `formatlist` context
  master_password    = "asdf"
}
```

### Debug Output
https://gist.github.com/rkoval/34d2b735e8668a3f7588159ea961694f

### Crash Output
N/A

### Expected Behavior
`data` interpolation in `formatlist` should return the appropriate `list` type

### Actual Behavior
`data` interpolation caused the resultant type of `formatlist` to _not_ be a `list`

### Steps to Reproduce
In both code snippets, run:
```
terraform plan -var test=hello
```

### Important Factoids
Running on macOS High Sierra, but nothing else that I can think of

### References
N/A

**Comment by neurolabs** [comment]
I am experiencing the same error when doing:

```
data "google_iam_policy" "foo" {
  binding {
    role = "roles/owner"
    members = "${list(
      data.terraform_remote_state.other.string_output, 
      "bar"
    )}"
}
```
removing the data... line from the definition makes it work.

**Comment by neurolabs** [comment]
even an indirection via locals does not work, if, at the end, something causes data interpolation or references data resources itself. Without data resources, it works.

**Comment by neurolabs** [comment]
still happening on terraform 0.11.1.

Up until now, we were able to work around this problem, but not anymore.

Are you aware of the problem? Do you understand that even an indirection via locals does not work?

Can you please reprioritize or at least tell us about priorities on this issue?

Thanks.

**Comment by apparentlymart** [comment]
Hi @neurolabs! Sorry for the issue and for an apparent regression in 0.11.1.

Can you provide a little more detail on what you were doing to work around this prior to 0.11.1 and what's breaking that workaround now? We didn't intentionally change anything here, but the failing workaround may give a clue to what's causing this bug.

We've not worked to address this issue specifically because we suspect it's part of a family of type-system related issues that we're going to address with our current project to rework the configuration language interpreter, but if we can get to the bottom of what changed in 0.11.1 we can hopefully at least restore your former workaround until the deeper configuration language improvement work is ready for release.

**Comment by apparentlymart** [comment]
This seems very similar to #16255. May be the same issue, or at least seems closely related.

**Comment by neurolabs** [comment]
Hi @apparentlymart . Thanks for the kind and informative reply. Let me serve you in return ;)

Sorry for the misunderstanding, but us being able to work around this issue does not involve any regression on your side, but functional requirements on our side. We need to do gcloud iam provisioning. And we need to do that in a big project, so there is a lot of iam stuff to handle. All of that needs to be defined in a single directory/bunch of tf-files, because otherwise, the definitions overwrite/cannibalize each other. We're assigning roles to

* service accounts via remote state files
* user accounts via variables

And some roles need to be assigned to a mix of service accounts and user accounts. Before, we were able to only assign specific roles to ONLY service accounts OR user accounts, OR to assign to static lists of a mix on all our cloud projects.

And we're at the tipping point where we need to join project specific lists of variables (user emails) with lists of data (remote state outputs, service account emails) in order to keep our iam structure working. This is not possible due to this issue.

Do you have a time frame for the rework to the interpreter? Can we support somehow?

**Comment by charles-at-geospock** [comment]
This feels very similar to the interpolation issue I have just reported as https://github.com/terraform-providers/terraform-provider-aws/issues/3125.

In that issue, the error report is 'blah is not a list', when what it means seems to be 'I expected a list but I got a value that cannot be computed until applied as part of my list elements'. Which seems similar here - in both cases the required value is a deferred interpolation which causes an unrelated message about the type of the value being wrong to be generated.

**Comment by apparentlymart** [comment]
Hi all! Sorry for the long silence here.

Based on the original issue and subsequent discussion, it looks like this is a different manifestation of the same root bug as #12263, which I've verified as fixed in v0.12.0-alpha1. This fix is merged in master and will be included in the forthcoming v0.12.0 final release. Thanks for reporting this bug, and thanks for your patience while we laid the groundwork to address it.
